packet/bgp/bgp.go

```diff
@@ -2489,32 +2489,139 @@ func flowSpecIpProtoParser(rf RouteFamily, args []string) (FlowSpecComponentInte
 }
 
 func flowSpecTcpFlagParser(rf RouteFamily, args []string) (FlowSpecComponentInterface, error) {
-	ss := make([]string, 0, len(TCPFlagNameMap))
-	for _, v := range TCPFlagNameMap {
-		ss = append(ss, v)
-	}
-	protos := strings.Join(ss, "|")
-	exp := regexp.MustCompile(fmt.Sprintf("^%s (not )?(match )?((((%s)\\&)*(%s) )*(((%s)\\&)*(%s)))$", FlowSpecNameMap[FLOW_SPEC_TYPE_TCP_FLAG], protos, protos, protos, protos))
-	elems := exp.FindStringSubmatch(strings.Join(args, " "))
-	if len(elems) < 1 {
-		return nil, fmt.Errorf("invalid flag format")
+	args = append(args[:0], args[1:]...) // removing tcp-flags string
+	fullCmd := strings.Join(args, " ")   // rebuiling tcp filters
+	err, flags, ops := parseTcpFlagCmd(fullCmd)
+	if err != nil {
+		return nil, err
 	}
 	items := make([]*FlowSpecComponentItem, 0)
-	op := 0
-	if elems[2] != "" {
-		op |= 0x1
+	for index := 0; index < len(flags); index++ {
+		items = append(items, NewFlowSpecComponentItem(ops[index], flags[index]))
 	}
-	if elems[1] != "" {
-		op |= 0x2
+	return NewFlowSpecComponent(FLOW_SPEC_TYPE_TCP_FLAG, items), nil
+}
+
+func parseTcpFlagCmd(myCmd string) (error, []int, []int) {
```

Comment:
yep makes sense ... I'll update the code according to your advices. Thanks for getting through it. I'll push a new version today of the week after as I am getting off for vacation next week.
Matt.